src/XrdOuc/XrdOucUtils.cc

```diff
@@ -1427,7 +1443,7 @@ static bool is_token_character(int c)
   if (isalnum(c))
     return true;
 
-  static constexpr char token_chars[] = "-._~+/=:";
+  static constexpr char token_chars[] = "-._~+/=:,";
```

Comment:
@bbockelm Sorry, I missed the explanation in the description. However, we cannot add comma here. Headers should be in their own lines, I don't see why adding comma is necessary.

Reply:
It's necessary because redaction isn't redacting only headers, it's also redacting CGI where tokens may show up.

So, if you have:

```
Authorization: foo
Authorization: bar
```

then the corresponding cgi is:

```
authz=foo,bar
```

Without this change, the logging is:

```
authz=REDACTED,bar
```

leaking the token `bar` into the log.  With the change, you have:

```
authz=REDACTED
```

Reply:
Yes, correct. But the problem is for the redaction of log lines, sometimes we see message like the one you saw in the unit test: `"(message: kXR_stat (path: /tmp/xrootd/public/foo?pelican.timeout=3s&authz=REDACTED, flags: none) ).",`
 which means that if we obfuscate the full line considering `,` is a valid token characeter, then we lose some information on the logs (`flags: none`).

Reply:
I think the conclusion from looking at the RFCs is that `Authorization:` should not appear more than once. Adding comma as valid token character is not great. How to handle these non-standard `authz` cgi elements in any case?

Reply:
> How to handle these non-standard authz cgi elements in any case?

Indeed, that's part of the next step and why I was working on #2381 in the first place.  For Pelican, we need to authenticate the cache and the client separately -- ensuring that an origin can ensure that (a) this is a valid request on behalf of a client and (b) the data is going to a known cache.  This prevents unregistered caches (or rogue clients) from accessing the origin directly.

That work is still on a branch (https://github.com/bbockelm/xrootd/tree/multi_token_authz) that isn't quite functional enough for even a draft -- however, getting both tokens to the plugin is a prerequisite.

Reply:
For caches, I think it's better to treat them like proxies, and use a [`Proxy-Authorization`](https://datatracker.ietf.org/doc/html/rfc7235#section-4.4) header rather than repeat the `Authorization:` header, which is not conforming to the standards. Or you can create `X-Pelican*` headers which can handle this outside of XRootD. I'd like to hear @abh3's and @osschar's opinions on this, but I hope they will agree that we want to avoid having non-standard behavior in XRootD, so that it can interoperate well with other software.

Reply:
I agree with @amadio that we really should avoid non-standard use of headers , especially when we are co-opting a standard header to do our bidding. In this particular case, I recall @bbockelm and I had a discussion on how to solve this particular problem by using RFC 7800 to generate special tokens that Xcache could use to access origins on behalf of clients (assuming you trust Xcache to do a proper client credential check - which we already do for x509). It appears that seems to have veered off course likely because I didn't produce the proposal to go that way. My other problem with multi-token transactions is that the framing overhead is moving in the wrong direction. We've already had problems with excessively large tokens and needed to make some code changes to accommodate them, This just makes tokens larger by compounding them. That said, if we need this kind of flexibility and if it is specific to Pelican, then either a special X-Pelican header or more generic special header, if this is useful beyond Pelican, would be warranted here. That doesn't address the obfuscation problem but I've already had a discussion with @amadio on a different mechanism of obfuscation that is insensitive to the token character set, which should drive the final nail into that coffin.